### ISSUE TITLE
New notebook: Neuropixel: Accessing Local Field potential. Visualizing LFP responses to a trial event, simple example with latencies, plot responses in different epochs.

### ISSUE THREAD

**Issue by jeromelecoq** [AllenInstitute/openscope_databook#19]
Carter will use both visual coding LFP NWB and Glow NWB files with LFP data to plot the averaged response to gratings for a number of repetition of the stimuli. There should be a response at the onset of the stimuli in the LFP data. 

It is worth talking with Jake about the rationale for this analysis. And this will also allow to debug https://github.com/AllenInstitute/OpenScopeNWB/issues/46

**Comment by rcpeene** [comment]
@Ahad-Allen does this mean I should wait on LFP timestamp analysis to continue progress on this script? Interpolation may not be necessary.

**Comment by Ahad-Allen** [comment]
Hmm, that might be correct. I'll have an updated file for you either the day of today or early tomorrow to look at.

**Comment by jeromelecoq** [comment]
We discussed current state of the notebook and Carter will : 
- Add some doc
- Add an image visualization
- Add the standard deviation intervals. 
- Make a PR

Jerome will then review and we will share with Hyeyoung and Jake for feedbacks.

**Comment by jeromelecoq** [comment]
- We will send update LFP NWB files to Jake once Carter has approved the newer files. 
- Ahad will upload the Dandi once its approved
- Ahad will write an email to Jake summarizing the change along with an introduction on the notebooks from Carter
- We need to ask him WHY we need 4 times larger NWB files because this is a big ask.

**Comment by Ahad-Allen** [comment]
@rcpeene  and I looked over these files and they seemed good! The upload failed on Friday because of an error I had with the ecephys version of auto uploading. I will retrigger the upload with some changes and then email Jake.

**Comment by jeromelecoq** [comment]
@rcpeene will make a PR with all views and submit for our reviews.

**Comment by rcpeene** [comment]
PR For LFP visualization notebook on #59. After @Ahad-Allen or @jeromelecoq review it, I will push it and subsequently ask Hyeyoung for feedback.

**Comment by jeromelecoq** [comment]
So, I had a lot of issues with running the environment for this notebook on macOS Monterey (12.5.1). 
I did the following: 
conda create --name databook_openscope_v1 python=3.8

conda activate databook_openscope_v1   

pip install -r requirements.txt 

But the environment could not complete. 
First, it could not find its specific version. I removed the version tag, but it could find itk at all. 

Then I commented itk-core and all itk packages. The install wen further. But then it had issues with tables. 
There it could find hdf5, so I then did
conda install hdf5
This fixed it

But then it could not find "vtk". Exact some problem as itk. I removed tag version and again it could not find the package itk at all. 
When I commented vtk, I got a lot of dependencies issues from packages that requires vtk and stopped there. 

I suspect we need a strategy to make this book deploy on multiple OS 

I will use binder for testing now.

**Comment by jeromelecoq** [comment]
Notebook launches on Binder :-)

**Comment by jeromelecoq** [comment]
Specific comment: 
This cell: 
stim_filepath = "your/filepath/spike_times.nwb"
lfp_filepath = "your/filepath/probeA_lfp.nwb"

All our notebooks should work without edits in my opinion. So I recommend modifying to download the files used for the notebook.

**Comment by jeromelecoq** [comment]
Also, I think we should have all cell output committed to the PR so that we can see the notebook without execution.

**Comment by jeromelecoq** [comment]
There are a couple typos in the text " it is recommended that write code to iterate through this table and filter all but the rows that contain an important stimulus event. The output should be a list of timestamps."

**Comment by jeromelecoq** [comment]
The interpolation scheme is a bit specific to strictly kHz rate. I think that would not be general purpose. I recommend looking into making this code compatible with sampling rates that are not powers of 10.

**Comment by rcpeene** [comment]
@jeromelecoq What do you mean by making it compatible? Are you referring to the fact that it is interpolated to a time axis that has units of 1 millisecond? A user can easily change this to be whatever units they want. Is that sufficient?

**Comment by jeromelecoq** [comment]
@rcpeene if the LFP timestamps were acquired a 50kHz, would your code work? I think it is currently converting timestamps into indices by truncating with int() casting. This part is perhaps not compatible and is implementing some kind of nearest neighbor interpolation in the background. Is that right?

**Comment by rcpeene** [comment]
Ah yes, I did consider this. Thanks for reminding me. The int() casting with the 1000 coefficient is the most efficient way to convert the time unit to the index within the interpolated data. If the user wanted to use a different unit for the interpolated data, they could simply change the 1000 within the int() casting to be something else.

This is not clear to the user though. I can make adjustments to make the interpolated units a clearly described variable which only needs to be changed in one place.

**Comment by rcpeene** [comment]
Tagging @westerberg-science for feedback.
Related PR #59

**Comment by jeromelecoq** [comment]
@Ahad-Allen will help with pointing at the right file for LFP

**Comment by jeromelecoq** [comment]
@rcpeene will commit a notebook with the output image of the cells to help @westerberg-science visualize the notebook

**Comment by jeromelecoq** [comment]
notes from Jake, copied from @rcpeene :

- potentials (mV) along depth. Maybe make the plot more compact. Axis need to be labeled. 
- catch-all category - extracting all stimulus. 
- comparing resting state with stimulus: Will be protocol specific. 
- Two ways to tackle this: One is to have a portion of the experiment with stimuli. 
The other is shuffle frames. 
- perhaps drawing all raw traces alongside average. Nice to have, yes. 
- power spectral density, we should consider this. 

**Priority:** 
- perhaps drawing all raw traces alongside average. Nice to have, yes. 
- potentials (mV) along depth. Maybe make the plot more compact. Axis need to be labeled. 
- Using visual coding data
- comparing resting state with stimulus: Will be protocol specific. 
For this, the visual coding should have a period of no stim to compare. 

- power spectral density, we should consider this. 
This is another notebook maybe

**Comment by jeromelecoq** [comment]
I added some comments on the PR

**Comment by jeromelecoq** [comment]
I am not sure how the depth issue was addressed. Perhaps just give the probe index to the label when you create the colorbar for the depth plot.

**Comment by jeromelecoq** [comment]
The only question is about this increasing variance with depth. I will check on binder, playing with the code.

**Comment by jeromelecoq** [comment]
Running the notebook on binder. The path was wrong. With Carter's help. I had to edit to download_loc = "../../data/"

**Comment by jeromelecoq** [comment]
[3:45 PM] Jerome Lecoq
The kernel for docs/visualization/visualize_lfp_responses.ipynb appears to have died. It will restart automatically.

[3:45 PM] Jerome Lecoq
memory?

[3:46 PM] Carter Peene
That is weird

[3:46 PM] Carter Peene
Could be memory

[3:46 PM] Carter Peene
which cell was it?

[3:46 PM] Carter Peene
perhaps the interpolation step

[3:48 PM] Jerome Lecoq
# generate regularly-space x values and interpolate along it
time_axis = np.arange(0, lfp.timestamps[-1], step=(1/interp_hz))
f = interpolate.interp1d(lfp.timestamps, lfp.data, axis=0, kind="nearest", fill_value="extrapolate")
interp_lfp = f(time_axis)

print(interp_lfp.shape)

[3:48 PM] Jerome Lecoq
yes this one

[3:48 PM] Jerome Lecoq
I saw at the bottom that Mem touched 2048 MB

[3:48 PM] Jerome Lecoq
it was red

**Comment by jeromelecoq** [comment]
@rcpeene will adapt the notebook to work with 2GB of RAM for now

**Comment by rcpeene** [comment]
Discussion for the adaptation in issue #82
